crates/rooch-rpc-server/src/server/rooch_server.rs

```diff
@@ -163,54 +162,37 @@ impl RoochAPIServer for RoochServer {
         limit: Option<StrView<u64>>,
         event_options: Option<EventOptions>,
     ) -> RpcResult<EventPageView> {
-        let (event_handle_id, _sender, last_event_seq) = self
-            .aggregate_service
-            .get_event_handle(event_handle_type.clone().into())
-            .await?;
-
         let event_options = event_options.unwrap_or_default();
         let cursor = cursor.map(|v| v.0);
         let limit = limit.map(|v| v.0);
+
         // NOTE: fetch one more object to check if there is next page
         let limit_of = min(limit.unwrap_or(DEFAULT_RESULT_LIMIT), MAX_RESULT_LIMIT);
-
-        let start = cursor.unwrap_or(0);
-        let end = min(start + (limit_of + 1), last_event_seq + 1);
-        let event_seqs: Vec<_> = if cursor.is_some() {
-            ((start + 1)..=end).collect()
-        } else {
-            (start..end).collect()
-        };
-        let event_ids = event_seqs
-            .into_iter()
-            .map(|v| EventID::new(event_handle_id, v))
-            .collect::<Vec<_>>();
-
+        let limit = limit_of + 1;
         let mut data = if event_options.decode {
             self.rpc_service
-                .get_events_by_event_ids(event_ids)
+                .get_events_by_event_handle(event_handle_type.into(), cursor, limit)
                 .await?
                 .into_iter()
-                .flatten()
                 .map(EventView::from)
                 .collect::<Vec<_>>()
         } else {
             //TODO provide a function to directly get the Event, not the AnnotatedEvent
             self.rpc_service
-                .get_events_by_event_ids(event_ids)
+                .get_events_by_event_handle(event_handle_type.into(), cursor, limit)
                 .await?
                 .into_iter()
-                .flatten()
                 .map(|e| EventView::from(e.event))
                 .collect::<Vec<_>>()
         };
 
         let has_next_page = (data.len() as u64) > limit_of;
-        data.truncate(limit_of as usize);
         let next_cursor = data
             .last()
             .map_or(cursor, |event| Some(event.event_id.event_seq));
 
+        data.truncate(limit_of as usize);
```

Comment:
In the previously implemented version, when the cursor is not None, the cursor is skipped the next time the page is turned and returned from the cursor+1 position. This semantic has changed here.

Reply:
As the testsuite case of this version:

1. cursor:0(None), limit:1, result: Event{event_seq:0}, next_cursor: 1
2. cursor:1, limit:1, result: Event{event_seq:1}, next_cursor: 2

If the previous version, should it be like the following?

1. cursor:-1(None), limit:1, result: Event{event_seq:0}, next_cursor: 0
2. cursor:0, limit:1, result: Event{event_seq:1}, next_cursor: 1

If the last item of the result is the same as the next_cursor, the next_cursor seems useless.

Reply:
Now the semantics of `next_cursor` is exactly the cursor that currently returns the result "the last item of the current page".

The earliest version used your method, and there was a problem:

When `has_next_page` is true, the normal `next_cursor` is returned; when `has_next_page` is false, the cursor corresponding to the last item of data is returned, which requires special processing.

For example, the following event only has 2 items of data:

1. cursor:-1(None), limit:1, result: Event{event_seq:0}, next_cursor: 1
2. cursor:1, limit:1, result: Event{event_seq:1}, next_cursor: 1

The previous DC discussion is here, https://discord.com/channels/1078938449974935592/1117770319436984341/1118013965189001216

Reply:
We need to discuss this, and I will refactor this later.